src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,596 @@
+                     *       and store the passed in Downstream<? super RR>
+                     *       downstream as an instance field in integrate()
+                     *       and read it in push(R r).
+                     *    2. Allocate a new Gatherer.Downstream<? super R> for
+                     *       each invocation of integrate() which might prove
+                     *       costly.
+                     */
+                    public boolean rightIntegrate(R r, Downstream<? super RR> downstream) {
+                        // The following logic is highly performance sensitive
+                        return (rightGreedy || rightProceed)
+                                && (rightIntegrator.integrate(rightState, r, downstream)
+                                || rightGreedy
+                                || (rightProceed = false));
+                    }
+                }
+
+                return new GathererImpl<T, State, RR>(
+                        State::new,
+                        (leftGreedy && rightGreedy)
+                                ? Integrator.<State, T, RR>ofGreedy(State::integrate)
+                                : Integrator.<State, T, RR>of(State::integrate),
+                        (leftCombiner == Gatherer.defaultCombiner()
+                        || rightCombiner == Gatherer.defaultCombiner())
+                                ? Gatherer.defaultCombiner()
+                                : State::joinLeft,
+                        (leftFinisher == Gatherer.<A, R>defaultFinisher()
+                        && rightFinisher == Gatherer.<AA, RR>defaultFinisher())
+                                ? Gatherer.defaultFinisher()
+                                : State::finish
+                );
+            }
+        }
+    }
+
+    // Public built-in Gatherers and factory methods for them
+
+    /**
+     * Gathers elements into fixed-size windows. The last window may contain
+     * fewer elements than the supplied window size.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: [[1, 2, 3], [4, 5, 6], [7, 8]]
+     * List<List<Integer>> windows =
+     *     Stream.of(1,2,3,4,5,6,7,8).gather(Gatherers.windowFixed(3)).toList();
+     * }
+     *
+     * @param windowSize the size of the windows
+     * @param <TR> the type of elements the returned gatherer consumes
+     *             and the contents of the windows it produces
+     * @return a new gatherer which groups elements into fixed-size windows
+     * @throws IllegalArgumentException when groupSize is less than 1
+     */
+    public static <TR> Gatherer<TR, ?, List<TR>> windowFixed(int windowSize) {
+        if (windowSize < 1)
+            throw new IllegalArgumentException("'groupSize' must be greater than zero");
+
+        return Gatherer.ofSequential(
+                // Initializer
+                () -> new ArrayList<>(windowSize),
```

Comment:
Note that if `windowSize` happens to be much bigger than the stream size (e.g., 1_000_000_000), we will waste a lot of memory and may fail with OutOfMemoryError for no reason, instead of collecting the whole stream into a single window. Not sure whether we should do anything with this. Probably using `SpinedBuffer` and limiting the initial size, like `Math.min(windowSize, 1_000_000)` will be better. WDYT?